logger.go

```diff
@@ -0,0 +1,155 @@
+package coraza
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/corazawaf/coraza/v3/debuglog"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type logger struct {
+	*zap.Logger
+	level debuglog.Level
+}
+
+var _ debuglog.Logger = (*logger)(nil)
+
+func newLogger(l *zap.Logger) debuglog.Logger {
+	return &logger{
+		Logger: l,
+		level:  debuglog.LevelInfo,
+	}
+}
+
+func (l *logger) WithOutput(w io.Writer) debuglog.Logger {
+	return &logger{
+		Logger: l.Logger.WithOptions(
+			zap.WrapCore(func(c zapcore.Core) zapcore.Core {
+				return zapcore.NewCore(
+					zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()),
+					zapcore.AddSync(w),
+					l.Logger.Level(),
+				)
+			}),
+		),
+		level: l.level,
+	}
+}
+
+func (l *logger) WithLevel(level debuglog.Level) debuglog.Logger {
+	return &logger{l.Logger, level}
+}
+
+func (l *logger) With(fields ...debuglog.ContextField) debuglog.Logger {
+	if len(fields) == 0 {
+		return l
+	}
+
+	defaultEvt := event{}
+	for _, f := range fields {
+		f(&defaultEvt)
+	}
+
+	return &logger{
+		Logger: l.Logger.With(defaultEvt.fields...),
+		level:  l.level,
+	}
+}
+
+func (l *logger) Trace() debuglog.Event {
+	if l.level < debuglog.LevelTrace {
+		return noopEvent{}
+	}
+
+	return &event{logger: l.Logger.Debug}
+}
+
+func (l *logger) Debug() debuglog.Event {
+	if l.level < debuglog.LevelDebug {
+		return noopEvent{}
+	}
+
+	return &event{logger: l.Logger.Debug}
+}
+
+func (l *logger) Info() debuglog.Event {
+	if l.level < debuglog.LevelInfo {
+		return noopEvent{}
+	}
+
+	return &event{logger: l.Logger.Info}
+}
+
+func (l *logger) Warn() debuglog.Event {
+	if l.level < debuglog.LevelWarn {
+		return noopEvent{}
+	}
+
+	return &event{logger: l.Logger.Warn}
+}
+
+func (l *logger) Error() debuglog.Event {
+	if l.level < debuglog.LevelError {
+		return noopEvent{}
+	}
+
+	return &event{logger: l.Logger.Error}
+}
+
+type event struct {
+	logger func(msg string, fields ...zap.Field)
+	fields []zap.Field
+}
+
+var _ debuglog.Event = (*event)(nil)
+
+func (e *event) Msg(msg string) {
+	e.logger(msg, e.fields...)
+}
+
+func (e *event) Str(key, val string) debuglog.Event {
+	e.fields = append(e.fields, zap.String(key, val))
+	return e
+}
+
+func (e *event) Bool(key string, b bool) debuglog.Event {
+	e.fields = append(e.fields, zap.Bool(key, b))
+	return e
+}
+
+func (e *event) Int(key string, i int) debuglog.Event {
+	e.fields = append(e.fields, zap.Int(key, i))
+	return e
+}
+
+func (e *event) Uint(key string, i uint) debuglog.Event {
+	e.fields = append(e.fields, zap.Uint(key, i))
+	return e
+}
+
+func (e *event) Stringer(key string, val fmt.Stringer) debuglog.Event {
+	e.fields = append(e.fields, zap.Stringer(key, val))
+	return e
+}
+
+func (e *event) Err(err error) debuglog.Event {
+	e.fields = append(e.fields, zap.Error(err))
+	return e
+}
+
+func (e *event) IsEnabled() bool {
+	return true
+}
+
+type noopEvent struct{}
```

Comment:
In the end I think it is worth to export the noop event, maybe as a variable or just the constructor @anuraaga

Reply:
It seems this is because of the level check in this code - I wonder if level checks are supposed to happen at a higher level in coraza (these become just LogWriter or something). Maybe too much to change now.

Another option could be to delegate to noop logger, which is public, instead of returning noopEvent - what about that?

Reply:
Checks will happen if we need to adapt logger with an underlying platform logger (in this case zap). I will leave this as it is now and if in the future we need the noop even to be exported we can export.